Add modbus light brightness and color temperature

DioSWolF opened the pull request:
This pull request introduces support for brightness and color temperature adjustments through switches. Previously, these functionalities were unavailable or required alternative implementations. Now, users can control brightness and color temperature directly using dedicated switches.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added brightness control support.
Implemented color temperature adjustments.
Updated relevant logic to integrate these features with existing switch mechanisms. Ensured compatibility with the current Modbus setup in Home Assistant. Testing:
Verified brightness adjustments work correctly through the switch. Tested color temperature changes to ensure seamless integration. Checked backward compatibility to avoid breaking existing functionality. Let me know if you need any modifications!

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37973
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

DioSWolF committed:
Add Support for Brightness and Color Temperature via Switches

This pull request introduces support for brightness and color temperature adjustments through switches. Previously, these functionalities were unavailable or required alternative implementations. Now, users can control brightness and color temperature directly using dedicated switches.

Changes:
Added brightness control support.
Implemented color temperature adjustments.
Updated relevant logic to integrate these features with existing switch mechanisms.
Ensured compatibility with the current Modbus setup in Home Assistant.
Testing:
Verified brightness adjustments work correctly through the switch.
Tested color temperature changes to ensure seamless integration.
Checked backward compatibility to avoid breaking existing functionality.
Let me know if you need any modifications!

DioSWolF committed:
Merge branch 'dev' into modbus_light

DioSWolF committed:
Update light.py

removed logger variable

crug80 reviewed (changes requested):
Juyst a quick and preliminary review before to check the code: 

1. The documentation for the new feature is totally missing. You have to write a PR and link it
2. Tests must be added

DioSWolF committed:
added tests for brightness and color temperature

DioSWolF commented:
I have some issues with documentation. Here is documentation for new functions  but i don`t know how to find and update old modbus documentation for light

[README.md](https://github.com/user-attachments/files/19178949/README.md)

crug80 commented:
> I have some issues with documentation. Here is documentation for new functions but i don`t know how to find and update old modbus documentation for light
> 
> [README.md](https://github.com/user-attachments/files/19178949/README.md)

You have to fork the next branch of the home-assistant.io repo and make a PR, then you can link the PR to this one.

crug80 commented:
Moreoever.. format the code following ruff guidelines.

DioSWolF committed:
formated code with ruff

DioSWolF commented:
here is PR for documentation [new documentation](https://github.com/home-assistant/home-assistant.io/pull/37971)

DioSWolF committed:
formated with ruff

crug80 commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -46,6 +46,8 @@
     CONF_UNIT_OF_MEASUREMENT,
     SERVICE_RELOAD,
+    CONF_BRIGHTNESS,
+    CONF_COLOR_TEMP,
```
Wrong ruff format

crug80 commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -226,6 +228,10 @@
+        vol.Optional(CONF_BRIGHTNESS): cv.positive_int,
+        vol.Optional(CONF_COLOR_TEMP): cv.positive_int,
+        vol.Optional(CONF_MIN_TEMP): cv.positive_int,
+        vol.Optional(CONF_MAX_TEMP): cv.positive_int,
```
Those lines should be aplied to the LIGHT_SCHEMA, not the switch

crug80 commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +51,161 @@ async def async_setup_platform(
+        self, hass: HomeAssistant, hub: ModbusHub, config: dict[str, Any]
+    ) -> None:
+        super().__init__(hass, hub, config)
+        self._state: bool = False
```
Why are using a local _state variable? The state is managed by the base_switch componenet. this should be removed.

crug80 reviewed (changes requested):
Please fix the code and use the pre-commit test to sanitaze other possible errors. then we'll review also the other part of the PR.

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +51,161 @@ async def async_setup_platform(
+        self._state: bool = False
+        self._brightness: int = DEFAULT_BRIGHTNESS
+        self._color_temp_kelvin: int | None = None
+        self.config = config
```
junior logic 😰

DioSWolF committed:
fix and sanitaze other possible errors in the code

DioSWolF committed:
Merge branch 'dev' into modbus_light

DioSWolF commented:
Hi! I missed something for review or all is good?

crug80 commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +354,60 @@ async def test_no_discovery_info_light(
+        },
+    ],
+)
+async def test_brightness_color_temp_control(hass: HomeAssistant, mock_modbus) -> None:
```
This is failing. 
Moreover, the coverage of the light.py is lower than 93%. Generally, all the added lines should be covered by tests.

DioSWolF replied:
i hope new tests will be better. Now Im understood a little bit how to create them

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +45,151 @@ async def async_setup_platform(
+
+    async def _async_update(self) -> None:
+        await super()._async_update()
+        """Update the entity state, including brightness and color temperature."""
```
if the issue was woti documentation, it was fixed last time

crug80 replied:
The issue is that ruff check has to pass.. Did you run the pre-commit test before submitting the PR?

DioSWolF committed:
add tests for color temperature, brightness, value convectors

DioSWolF committed:
change variable names, fixed convertor for color temperature, add check for brightness_result and color_result,

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
Update class comment

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF commented on homeassistant/components/modbus/const.py:
```diff
@@ -167,3 +168,12 @@ class DataType(str, Enum):
+DEFAULT_BRIGHTNESS = 255
+MIN_BRIGHTNESS = 0
+MAX_BRIGHTNESS = 255
+MODBUS_SCALE_MIN = 0
+MODBUS_SCALE_MAX = 100
```
this is not only temp scale, its a system with minimal and maximal value scale for all values

abmantis replied:
I don't see them being used anywhere else

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +45,150 @@ async def async_setup_platform(
+                use_call=CALL_TYPE_WRITE_REGISTER,
+            )
+
+            self._brightness = brightness
```
can you explaine why i need to set it when self._verify_active is False? If i understood correctly, verify config will not affected to the new functions

abmantis replied:
looking at the base Modbus Entity class, when `self._verify_active` is `True`, the state is fetched from the device, instead of being optimistically set directly as it is being done here.

DioSWolF committed:
rename functions

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF committed:
rename function

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +45,150 @@ async def async_setup_platform(
-        """Set light on."""
+        """Turn light on and set brightness if provided."""
         await self.async_turn(self.command_on)
+        await asyncio.sleep(0.01)
```
i will check again, but it is not possible to set brightness and color temp before command_on. 
remote sleep, was added from automatisation logic

DioSWolF commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -415,7 +417,14 @@
-LIGHT_SCHEMA = BASE_SWITCH_SCHEMA.extend({})
+LIGHT_SCHEMA = BASE_SWITCH_SCHEMA.extend(
+    {
+        vol.Optional(CONF_BRIGHTNESS): cv.positive_int,
+        vol.Optional(CONF_COLOR_TEMP): cv.positive_int,
```
will change documentation for this too

crug80 replied:
This PR is still ongoing so, you can make all the chages here.

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -30,11 +45,150 @@ async def async_setup_platform(
+    ) -> None:
+        """Initialize the Modbus light entity."""
+        super().__init__(hass, hub, config)
+        self._brightness: int = DEFAULT_BRIGHTNESS
```
> Should this be set to `None` if the light does not support brightness? Or be left at `None` until we get an update from the device on `_async_update`
I will check how it will work when  light does not support brightness. But for my logic, will be better to leave it like that. But if you will say change it, i will do it
we can`t take it from  `_async_update`, this is update from manual dimmers.

DioSWolF commented on homeassistant/components/modbus/const.py:
```diff
@@ -167,3 +168,12 @@ class DataType(str, Enum):
     (Platform.SWITCH, CONF_SWITCHES),
 )
+
+# Default values for light
```
thay are constants, i don't think it will be a good idea.

DioSWolF replied:
thank you! I will move them

DioSWolF committed:
remaned brightness and color temp registers variables

DioSWolF committed:
rename registers variables, color and brightness attr variables, function to convert percent to modbus, added modbus scale from min to max, check for answer without unavailible device,

DioSWolF committed:
rename function to convert percent to brightness

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
rename default variables

DioSWolF committed:
add verify checking

abmantis commented:
Hey @DioSWolF this is ready for review, right?

DioSWolF committed:
formated with ruff

DioSWolF committed:
change turn-on logic

abmantis commented:
Hey @DioSWolF , the CI is still failing. 
I really recommend that you setup the pre-commit hook or run it manually: https://developers.home-assistant.io/docs/development_testing/#preventing-linter-errors

DioSWolF commented:
![Screenshot_1](https://github.com/user-attachments/assets/e50574b9-da2c-4452-ae83-669d3875193b)
Here pre-commit results and it don`t have issues...

DioSWolF committed:
remove wrong imports

DioSWolF committed:
fixed mypy check errors

DioSWolF commented:
![image](https://github.com/user-attachments/assets/fd9b7a4b-427b-4fbe-ac6d-3fd8e6667f66)
![image](https://github.com/user-attachments/assets/207f360b-85d0-4891-9049-778a7b38ad0d)
Here local results

DioSWolF committed:
changed check for brightness and color temp

they can be 0

DioSWolF commented:
Before you take over PR, I learn a lot of things that I never noticed before in my work.😁

DioSWolF committed:
added return for pre-commit tests

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+        self._attr_color_temp_kelvin: int | None = None
+
+        # Determine color mode dynamically
+        self._attr_color_mode = self._detect_color_mode(config=config)
```
```suggestion
        self._attr_color_mode = self._detect_color_mode(config)
```
no advantage in using kwargs here

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+    async def async_set_brightness(self, brightness: int | Any) -> None:
+        """Set the brightness of the light."""
+        if self._brightness_address:
+            conv_brightness = await self._convert_brightness_to_modbus(
+                brightness=brightness
+            )
```
```suggestion
            conv_brightness = await self._convert_brightness_to_modbus(brightness)
```

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+                )
+
+    @staticmethod
+    async def _convert_modbus_percent_to_brightness(percent: int) -> int:
```
```suggestion
    def _convert_modbus_percent_to_brightness(percent: int) -> int:
```

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+            * LIGHT_MAX_BRIGHTNESS
+        )
+
+    async def _convert_modbus_percent_to_temperature(self, percent: int) -> int | None:
```
```suggestion
    def _convert_modbus_percent_to_temperature(self, percent: int) -> int | None:
```

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+        return None
+
+    @staticmethod
+    async def _convert_brightness_to_modbus(brightness: int) -> int:
```
```suggestion
    def _convert_brightness_to_modbus(brightness: int) -> int:
```

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+            * (LIGHT_MODBUS_SCALE_MAX - LIGHT_MODBUS_SCALE_MIN)
+        )
+
+    async def _convert_color_temp_to_modbus(self, kelvin: int) -> int | None:
```
```suggestion
    def _convert_color_temp_to_modbus(self, kelvin: int) -> int | None:
```

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+            conv_brightness = await self._convert_brightness_to_modbus(
+                brightness=brightness
+            )
+            if conv_brightness is not None:
```
this check is not needed since `_convert_brightness_to_modbus` returns `int` and not `None`

DioSWolF replied:
with new changes we will need it

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+
+    async def async_set_brightness(self, brightness: int | Any) -> None:
+        """Set the brightness of the light."""
+        if self._brightness_address:
```
nitpick, but we could reverse this and return early, to reduce scopes. like:

```python
if not self._brightness_address:
    return
conv_brightness = await self._convert_brightness_to_modbus(brightness)
...
```

DioSWolF replied:
can we leave it like it now? With this i need to create more tests to check this line 😁

abmantis replied:
any reason to revert this back?

abmantis replied:
same for color_temp

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+
+    async def _convert_modbus_percent_to_temperature(self, percent: int) -> int | None:
+        """Convert Modbus scale (0-100) to the color temperature in Kelvin (2000-7000 К)."""
+        if isinstance(self._attr_min_color_temp_kelvin, int) and isinstance(
+            self._attr_max_color_temp_kelvin, int
+        ):
```
lets reverse this if condition, log an error (because this should not happen) and `return None`.

DioSWolF replied:
@abmantis  if we will reverse it and add logger, we will need to add tests, but it will be unpissible to have them `None` in tests and have a call for that`s methods...How to be with it?

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +52,172 @@ async def async_setup_platform(
+            if (
+                brightness_result
+                and brightness_result.registers
+                and brightness_result.registers[0] != 0xFFFF
```
what does 0xFFFF mean for brigthness and color temp?

DioSWolF replied:
when device not ON but you try to read registers it return 65535 or 0xFFFF and we need to skip it

abmantis replied:
lets add a const for it then, so that reading the code makes it clear what it means. maybe something like MODBUS_INVALID_VALUE ?

DioSWolF replied:
@abmantis ready for review😊

DioSWolF committed:
Merge branch 'dev' into modbus_light

DioSWolF committed:
remove _attr_brightness and _attr_color_temp_kelvin from ModbusLight init, add check for  color temp and brightness before set it in device, verify checks, remove async fuunction for convectors

DioSWolF committed:
add logging, check for convertors inputs, color temp bounds

DioSWolF committed:
add tests for concertors, remove async convertors

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
add constant for invalid modbus value

DioSWolF committed:
Merge branch 'dev' into modbus_light

abmantis commented on tests/components/modbus/test_light.py:
should we also update `test_restore_state_light` to check that the new attributes are restored?

DioSWolF replied:
yes, we need and i will try to do it. But now `_async_update` rewrited values from device and tests failing...

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +384,199 @@ async def test_no_discovery_info_light(
+        }
+    ],
+)
+async def test_brightness_light(hass: HomeAssistant, mock_modbus_ha) -> None:
```
lets place this after `test_light_service_turn` since they are related

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +384,199 @@ async def test_no_discovery_info_light(
+    assert hass.states.get(ENTITY_ID).state == STATE_ON
+
+    calls = mock_modbus_ha.write_register.call_args_list
+    modbus_brightness = ModbusLight._convert_brightness_to_modbus(brightness=128)
```
lets use a constant value to avoid calling an internal method in the test

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +384,199 @@ async def test_no_discovery_info_light(
+        }
+    ],
+)
+async def test_color_temp_light(hass: HomeAssistant, mock_modbus_ha) -> None:
```
we could combine this test with the brightness one with test params. each param tuple would have both the service data for `turn_on` (which can include only temp, only brightness, or both) and the expected results.

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +384,199 @@ async def test_no_discovery_info_light(
+
+    calls = mock_modbus_ha.write_register.call_args_list
+    modbus_brightness = ModbusLight._convert_brightness_to_modbus(brightness=128)
+    modbus_temp = round(
```
lets use a constant value here too

DioSWolF committed:
remove not used variable

DioSWolF committed:
remove imput types and check for imput type from modbus device

DioSWolF committed:
reordering the code to reduce nested scopes

DioSWolF committed:
add variable for brightness

DioSWolF committed:
ruff formate

DioSWolF committed:
reordering the code to reduce nested scopes, add correct type for kwargs

DioSWolF committed:
add tests for bright and color temp convertors, checks for None values

Not sure then i did it correctly.. Please, give me feedback how to do correctly, i did`nt find other way for this

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF commented on tests/components/modbus/test_light.py:
```diff
@@ -352,3 +384,199 @@ async def test_no_discovery_info_light(
+
+
+@pytest.mark.asyncio
+class TestModbusConvectors:
```
@abmantis  created other tests, can you check it please, mb i do something wrong with them and i need some help, if yes...

abmantis replied:
![image](https://github.com/user-attachments/assets/94699456-bc69-4cbc-8ab0-f351525eaccf)
the code pushed to this branch still has this

DioSWolF replied:
you mean `config=config`...yep, my mistake

DioSWolF committed:
add restore states

DioSWolF committed:
add tests for restore brightness and color temp, move brightness  tests to the color temp test

DioSWolF committed:
remove kwarg name in func call

DioSWolF committed:
change kwargs type to Any

DioSWolF committed:
add type for hass in modbus_light_entity

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+        }
+    ],
+)
+async def test_color_temp_light(
```
```suggestion
async def test_color_temp_brightness_light(
```

abmantis replied:
can we also add an example for a light without brightness but only color temp?

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+        SERVICE_TURN_ON,
+        service_data={
+            ATTR_ENTITY_ID: ENTITY_ID,
+            ATTR_BRIGHTNESS: 128,
+            ATTR_COLOR_TEMP_KELVIN: 2000,
```
lets move the service attributes for bright and color to the parameters.
check https://github.com/home-assistant/core/blob/dev/tests/components/whirlpool/test_climate.py#L167 for an example.

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+        },
+        blocking=True,
+    )
+    await hass.async_block_till_done()
```
is this needed even with `blocking=True` above?

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+    await hass.async_block_till_done()
+    assert hass.states.get(ENTITY_ID).state == STATE_ON
+    calls = mock_modbus_ha.write_register.call_args_list
+    expected_modbus_brightness = 50
+    expected_modbus_temp = 0
```
lets also move this to the test params, since it should not be the same for both lights

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+    brightness = 255
+    modbus_color_temp = 0
+    incorrect_input = "invalid"
+    result = await modbus_light_entity.async_set_brightness(incorrect_input)
```
as mentioned in the comment above, lets not call entity methods directly but use service calls instead

DioSWolF replied:
if i will use service calls and try to add a str brightnes or color temp, i will have this error
![image](https://github.com/user-attachments/assets/fdd6b6b1-8bd1-4526-8962-552965ffdc0a)
I will check this link, but not sure then ii will helps with this issue
![image](https://github.com/user-attachments/assets/0cf83f2f-7522-4455-9558-1d16f700a65b)

abmantis replied:
maybe you are passing a string to the `color_temp_kelvin` attribute when calling the service?

DioSWolF replied:
@abmantis I’m not sure how to write proper tests for the checks involving '_attr_min_color_temp_kelvin', '_attr_max_color_temp_kelvin', and how set `color_temp_register` to `None`. Any guidance would be appreciated.

abmantis replied:
oh, change it so that `async_set_brightness` receives an `int` instead of `int | Any` (same for `_convert_brightness_to_modbus`). there is no need to check if it is an `int` since that is handled on the parent LightEntity class already

DioSWolF replied:
this wouldn't be a problem if it weren't for
![image](https://github.com/user-attachments/assets/06338d46-b89b-4d19-a65d-43b70184e6b1)

and one more issue with `_attr_min_color_temp_kelvin`, `_attr_max_color_temp_kelvin`, i can`t understand how to change them in the tests with service call...mb i can take it from `hass.data` change and rewrite again in `hass.data`, but now sure then it a good way

abmantis replied:
You need to cast the value to `int` before passing it to `async_set_brightness`. same for color temp.
For min/max color temp, can't you define that on the config? there is no need to change it during the test run. you can have different values as test parameters for example

DioSWolF replied:
No, i can`t set not int in config for `_attr_min_color_temp_kelvin`, `_attr_max_color_temp_kelvin` , but i can do something like this with them and with color address
![image](https://github.com/user-attachments/assets/b4126767-d3e4-4933-8772-568ee4781370)

DioSWolF replied:
![image](https://github.com/user-attachments/assets/9b8ba2bb-0842-41aa-b376-e28e1f414dec)
![image](https://github.com/user-attachments/assets/c9376917-337e-41c8-b3cc-5e4027d8d39d)

DioSWolF replied:
or i can chenge logic to get brightness and color temp params like this 
![image](https://github.com/user-attachments/assets/d0718e4d-a122-4cbe-b004-ee34c98c2b36)

DioSWolF replied:
[
5e987dc](https://github.com/home-assistant/core/pull/139703/commits/5e987dc6b13336780cc86d5ffc3fb4c3ee7cb7da)
https://github.com/home-assistant/core/pull/139703/commits/b37c9d421ac289f8776a325beec7b69b1548eda5
Here 2 changes with other tests system, what cover 100% code, and changes with types in light.py
I hope there all is good

abmantis replied:
Thanks for the changes! light.py looks good! 
I've left some comments on the tests.

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+    incorrect_input = "invalid"
+    result = await modbus_light_entity.async_set_brightness(incorrect_input)
+    assert result is None
+    modbus_light_entity._brightness_address = None
```
this should be part of the test parameterization, so that the test runs once for a light with brightness_address, and another for one without

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+    "do_config",
+    [
+        {
+            CONF_LIGHTS: [
```
we should still keep the old light config too

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -314,9 +338,101 @@ async def test_light_service_turn(
+                    CONF_BRIGHTNESS_REGISTER: 1,
+                    CONF_WRITE_TYPE: CALL_TYPE_COIL,
+                    CONF_VERIFY: {},
+                },
```
maybe I am wrong but it seems that this test is passing config for 2 lights, but only uses 1?

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -217,7 +223,18 @@ async def test_all_light(hass: HomeAssistant, mock_do_cycle, expected) -> None:
 @pytest.mark.parametrize(
     "mock_test_state",
-    [(State(ENTITY_ID, STATE_ON),)],
+    [
```
I think we should set config for 2 lights (a simple one and one with temp+brightness), and have different states for the 2 entities that get created for each config

DioSWolF committed:
remove default value for brightness

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF committed:
remove default value for color temp

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +54,202 @@ async def async_setup_platform(
+        if not self._verify_active:
+            return
+
+        if not self._brightness_address:
+            return
+
```
mb yes...i think about it and in my mind was -> Color temp can`t be without brightness, its mean if we don`t have brightness address we can return early
But mb that was mistake. I will return old check for brightness address

DioSWolF committed:
remove early return if we don`t have brightness

DioSWolF committed:
rename function

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

DioSWolF committed:
remove modbus_light_entity from pytest fixture, change tests  to use parameters, add test for no valid param

DioSWolF committed:
change types for inputs, change checks for input valies types

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
add 2 configs and statuses for restore statuses

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +54,192 @@ async def async_setup_platform(
+                    brightness_result.registers[0]
+                )
+
+        if not self._color_temp_address:
+            return
```
This is a nitpick and its fine if you prefer to keep it, but I would suggest reversing this condition and moving the color temp code to inside the if block.
I am very in favor of early returns, but since this method is not specific for the color temp, early returns like this make it easier to introduce a bug in the future if someone misses it and adds more code bellow.

DioSWolF replied:
i can change it, but in beginning i had color temp logic inside `if` and you told me to change it for early return😁

abmantis replied:
Eheh :D  That other suggestion to have an early-return was for the `async_set_brightness` method. Since that method has only 1 single purpose (brightness) it is fine to early return if something is missing.

The suggestion I made now for not having early-return was for `_async_update` which handles multiple different things, so it should (IMO) have different scopes for those different things. :sweat_smile:

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -307,21 +347,336 @@ async def test_light_service_turn(
+
+    service_data = {ATTR_ENTITY_ID: ENTITY_ID}
+
+    if brightness_input is not None:
+        service_data[ATTR_BRIGHTNESS] = brightness_input
+
+    if color_temp_input is not None:
+        service_data[ATTR_COLOR_TEMP_KELVIN] = color_temp_input
```
since we should not have conditions in tests (some old tests still have them, but newer ones shouldn't), lets put the whole `service_data` in the test params.
You can just set it as a object in the test params like 
```python
{ATTR_BRIGHTNESS: 128}
```
and
```python
{ATTR_BRIGHTNESS: 128, ATTR_COLOR_TEMP_KELVIN: 3000}
```

Then in the test body you can merge the param with the entity_id part:
```python
        {ATTR_ENTITY_ID: ENTITY_ID, **service_data},
```

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -307,21 +347,336 @@ async def test_light_service_turn(
+    )
+    assert hass.states.get(ENTITY_ID).state == STATE_ON
+    calls = mock_modbus_ha.write_register.call_args_list
+    if expected_brightness is not None:
+        assert any(
+            call.args[0] == 1 and call.kwargs["value"] == expected_brightness
+            for call in calls
+        )
+    if expected_color_temp is not None:
+        assert any(
+            call.args[0] == 2 and call.kwargs["value"] == expected_color_temp
+            for call in calls
+        )
```
could we also move these into the test params to remove the conditions? even if we need to remove the one without brightness or color temp

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -307,21 +347,336 @@ async def test_light_service_turn(
+
+    entities = list(hass.data["light"].entities)
+    entity = entities[0]
+    entity._attr_min_color_temp_kelvin = None
```
we should not change the internal state directly. Why not set `CONF_MIN_TEMP`/ `CONF_MAX_TEMP` in the config above?

DioSWolF replied:
and the same with `color_temp_address`, it not works when i use only `brightness_address` in config...

abmantis replied:
Right, because we validate the configs so None is never accepted as input.

Also, it is OK if coverage is not 100%. As long as it is [at least 95%, it is fine](https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/test-coverage/)! But let me try to get a second opinion on this.

DioSWolF replied:
we can remove `test_color_temp_no_valid_params`, what you don't like, and coverage will **be 96%**

abmantis replied:
I talked about this with another core member, and I think it is OK to just remove the `if` checks for _attr_min_color_temp_kelvin and _attr_max_color_temp_kelvin.
Correct me if I am wrong, but they should never be None or not-int when they are used, right? 

It will never reach that point in the code if the light does not have the corresponding registers. If it has the registers, we set the _attr_min_color_temp_kelvin/_attr_max_color_temp_kelvin. So it is safe to just remove the check.

DioSWolF replied:
yep, correctly. Ok, i will remove them and remove tests for that too

DioSWolF replied:
we will have check for that params, because MYPY tests filed without them...

abmantis replied:
since we know it should be `int` at that point, can you replace the `if` statement with an `assert` and simplify the return type? like:

```python
    def _convert_modbus_percent_to_temperature(self, percent: int) -> int:
        """Convert Modbus scale (0-100) to the color temperature in Kelvin (2000-7000 К)."""
        assert isinstance(self._attr_min_color_temp_kelvin, int) and isinstance(
            self._attr_max_color_temp_kelvin, int
        )
        return round(
            self._attr_min_color_temp_kelvin
            + (
                percent
                / (LIGHT_MODBUS_SCALE_MAX - LIGHT_MODBUS_SCALE_MIN)
                * (self._attr_max_color_temp_kelvin - self._attr_min_color_temp_kelvin)
            )
        )
```

DioSWolF replied:
good idea...i didn`t know then `assert` works like that

DioSWolF committed:
pylint fixes

DioSWolF committed:
move service data to the test params

DioSWolF committed:
move color temp and brightness logic inside if

DioSWolF committed:
format with ruff

DioSWolF committed:
change tests to use only color temp and brightness params

DioSWolF committed:
remove tests and logger for _attr_min_color_temp_kelvin and  _attr_max_color_temp_kelvin

DioSWolF committed:
Merge branch 'dev' into modbus_light

DioSWolF committed:
remove early return from _async_update

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
ruff format

DioSWolF committed:
fix address check

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -338,6 +495,29 @@ async def test_service_light_update(hass: HomeAssistant, mock_modbus_ha) -> None
+        },
+        blocking=True,
+    )
+    if hass.states.get(ENTITY_ID).attributes.get(ATTR_BRIGHTNESS):
+        assert (
+            hass.states.get(ENTITY_ID).attributes.get(ATTR_BRIGHTNESS)
+            == expected_brightness
```
```suggestion
    assert (expected_brightness is None or
        hass.states.get(ENTITY_ID).attributes.get(ATTR_BRIGHTNESS)
        == expected_brightness
```
that way we tie this to the test input and not to the state (which may be wrong)

abmantis replied:
same for the color temp bellow

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -307,21 +347,138 @@ async def test_light_service_turn(
+            {ATTR_BRIGHTNESS: 128, ATTR_COLOR_TEMP_KELVIN: 2000},
+            50,
+            0,
+        ),
+    ],
+)
+async def test_color_temp_brightness_light(
+    hass: HomeAssistant,
+    mock_modbus_ha,
+    service_data,
+    expected_brightness,
+    expected_color_temp,
+) -> None:
+    """Test Modbus Light color temperature and brightness."""
+    assert hass.states.get(ENTITY_ID).state == STATE_OFF
+    await hass.services.async_call(
+        LIGHT_DOMAIN,
+        SERVICE_TURN_ON,
+        service_data={ATTR_ENTITY_ID: ENTITY_ID, **service_data},
+        blocking=True,
+    )
+    assert hass.states.get(ENTITY_ID).state == STATE_ON
+    calls = mock_modbus_ha.write_register.call_args_list
+
+    assert any(
+        call.args[0] == 1 and call.kwargs["value"] == expected_brightness
+        for call in calls
+    )
+    assert any(
+        call.args[0] == 2 and call.kwargs["value"] == expected_color_temp
+        for call in calls
+    )
+
+    await hass.services.async_call(
+        LIGHT_DOMAIN,
+        SERVICE_TURN_OFF,
+        service_data={ATTR_ENTITY_ID: ENTITY_ID},
+        blocking=True,
+    )
+    assert hass.states.get(ENTITY_ID).state == STATE_OFF
```
```suggestion
        (
            {
                CONF_LIGHTS: [
                    {
                        CONF_NAME: TEST_ENTITY_NAME,
                        CONF_ADDRESS: 1234,
                        CONF_WRITE_TYPE: CALL_TYPE_REGISTER_HOLDING,
                        CONF_SCAN_INTERVAL: 0,
                        CONF_BRIGHTNESS_REGISTER: 1,
                        CONF_COLOR_TEMP_REGISTER: 2,
                    }
                ]
            },
            {ATTR_BRIGHTNESS: 128, ATTR_COLOR_TEMP_KELVIN: 2000},
            [(1, 50), (2, 0)],
        ),
        (
            {
                CONF_LIGHTS: [
                    {
                        CONF_NAME: TEST_ENTITY_NAME,
                        CONF_ADDRESS: 1234,
                        CONF_WRITE_TYPE: CALL_TYPE_REGISTER_HOLDING,
                        CONF_SCAN_INTERVAL: 0,
                        CONF_BRIGHTNESS_REGISTER: 1,
                    }
                ]
            },
            {ATTR_BRIGHTNESS: 256},
            [(1, 100)],
        ),
        (
            {
                CONF_LIGHTS: [
                    {
                        CONF_NAME: TEST_ENTITY_NAME,
                        CONF_ADDRESS: 1234,
                        CONF_WRITE_TYPE: CALL_TYPE_REGISTER_HOLDING,
                        CONF_SCAN_INTERVAL: 0,
                        CONF_COLOR_TEMP_REGISTER: 2,
                    }
                ]
            },
            {ATTR_BRIGHTNESS: 128, ATTR_COLOR_TEMP_KELVIN: 3000},
            [(2, 20)],
        ),
    ],
)
async def test_color_temp_brightness_light(
    hass: HomeAssistant,
    mock_modbus_ha,
    service_data,
    expected_calls,
) -> None:
    """Test Modbus Light color temperature and brightness."""
    assert hass.states.get(ENTITY_ID).state == STATE_OFF
    await hass.services.async_call(
        LIGHT_DOMAIN,
        SERVICE_TURN_ON,
        service_data={ATTR_ENTITY_ID: ENTITY_ID, **service_data},
        blocking=True,
    )
    assert hass.states.get(ENTITY_ID).state == STATE_ON
    calls = mock_modbus_ha.write_register.call_args_list

    for expected_register, expected_value in expected_calls:
        assert any(
            call.args[0] == expected_register and call.kwargs["value"] == expected_value
            for call in calls
        ), (
            f"Expected register {expected_register} with value {expected_value} not found in calls {calls}"
        )

    await hass.services.async_call(
        LIGHT_DOMAIN,
        SERVICE_TURN_OFF,
        service_data={ATTR_ENTITY_ID: ENTITY_ID},
        blocking=True,
    )
    assert hass.states.get(ENTITY_ID).state == STATE_OFF
```

we can make it so that the call args are also part of the test args, so we can have more variants of the light config

abmantis commented on tests/components/modbus/test_light.py:
```diff
@@ -307,21 +347,138 @@ async def test_light_service_turn(
 
 @pytest.mark.parametrize(
-    "do_config",
+    (
+        "do_config",
+        "service_data",
+        "expected_brightness",
+        "expected_color_temp",
+    ),
```
```suggestion
    ("do_config", "service_data", "expected_calls"),
```

(part of my suggestion for the method bellow, since github wouldn't let me suggest the whole thing in 1 comment)

DioSWolF committed:
return early returns in async_set_color_temp async_set_brightness

DioSWolF committed:
change if to assert in color temp convertors

DioSWolF committed:
change tests inputs, add iteration for sone tests and inputs

DioSWolF committed:
Merge branch 'dev' into modbus_light

abmantis commented on homeassistant/components/modbus/light.py:
```diff
@@ -32,9 +54,177 @@ async def async_setup_platform(
+
+        conv_color_temp_kelvin = self._convert_color_temp_to_modbus(color_temp_kelvin)
+
+        if conv_color_temp_kelvin is not None:
+            await self._hub.async_pb_call(
+                unit=self._slave,
+                address=self._color_temp_address,
+                value=conv_color_temp_kelvin,
+                use_call=CALL_TYPE_WRITE_REGISTER,
+            )
```
```suggestion
    await self._hub.async_pb_call(
        unit=self._slave,
        address=self._color_temp_address,
        value=conv_color_temp_kelvin,
        use_call=CALL_TYPE_WRITE_REGISTER,
    )
```

abmantis reviewed (changes requested):
Just one cleanup missing from the latest type changes and I think this is ready!

DioSWolF committed:
remove None check for color temp

DioSWolF committed:
Merge branch 'modbus_light' of https://github.com/DioSWolF/core_homeassistant into modbus_light

DioSWolF committed:
Merge branch 'dev' into modbus_light

abmantis reviewed (approved):
Thanks a lot @DioSWolF !

DioSWolF commented:
Thank you @abmantis !
